packetbeat/protos/cassandra/frame.go

```diff
@@ -0,0 +1,900 @@
+	r io.Reader
+
+	proto byte
+	// flags are for outgoing flags, enabling compression and tracing etc
+	flags    byte
+	compres  Compressor
+	headSize int
+	// if this frame was read then the header will be here
+	header *frameHeader
+
+	// if tracing flag is set this is not nil
+	traceID []byte
+
+	// holds a ref to the whole byte slice for rbuf so that it can be reset to
+	// 0 after a read.
+	readBuffer []byte
+
+	rbuf []byte
+}
+
+func newFramer(r io.Reader, compressor Compressor, version byte) *framer {
+	f := framerPool.Get().(*framer)
+	var flags byte
+	if compressor != nil {
+		flags |= flagCompress
+	}
+
+	version &= protoVersionMask
+
+	headSize := 8
+	if version > protoVersion2 {
+		headSize = 9
+	}
+
+	f.compres = compressor
+	f.proto = version
+	f.flags = flags
+	f.headSize = headSize
+
+	f.r = r
+	f.rbuf = f.readBuffer[:0]
+
+	f.header = nil
+	f.traceID = nil
+
+	return f
+}
+
+type frame interface {
+	Header() frameHeader
+}
+
+func readHeader(r io.Reader, p []byte) (head frameHeader, err error) {
+	_, err = io.ReadFull(r, p[:1])
+	if err != nil {
+		return frameHeader{}, err
+	}
+
+	version := p[0] & protoVersionMask
+
+	if version < protoVersion1 || version > protoVersion4 {
+		return frameHeader{}, fmt.Errorf("unsupported response version: %d", version)
+	}
+
+	headSize := 9
+	if version < protoVersion3 {
+		headSize = 8
+	}
+
+	_, err = io.ReadFull(r, p[1:headSize])
+	if err != nil {
+		return frameHeader{}, err
+	}
+
+	p = p[:headSize]
+
+	v := p[0]
+
+	head.version = protoVersion(v)
+
+	head.flags = p[1]
+
+	if version > protoVersion2 {
+		if len(p) != 9 {
+			return frameHeader{}, fmt.Errorf("not enough bytes to read header require 9 got: %d", len(p))
+		}
+
+		head.stream = int(int16(p[2])<<8 | int16(p[3]))
+		head.op = frameOp(p[4])
+		head.length = int(readInt(p[5:]))
+	} else {
+		if len(p) != 8 {
+			return frameHeader{}, fmt.Errorf("not enough bytes to read header require 8 got: %d", len(p))
+		}
+
+		head.stream = int(int8(p[2]))
+		head.op = frameOp(p[3])
+		head.length = int(readInt(p[4:]))
+	}
+
+	return head, nil
+}
+
+// reads a frame form the wire into the framers buffer
+func (f *framer) readFrame(head *frameHeader) error {
+	if head.length < 0 {
+		return fmt.Errorf("frame body length can not be less than 0: %d", head.length)
+	} else if head.length > maxFrameSize {
+		// need to free up the connection to be used again
+		_, err := io.CopyN(ioutil.Discard, f.r, int64(head.length))
+		if err != nil {
+			return fmt.Errorf("error whilst trying to discard frame with invalid length: %v", err)
+		}
+		return ErrFrameTooBig
+	}
+
+	if cap(f.readBuffer) >= head.length {
+		f.rbuf = f.readBuffer[:head.length]
+	} else {
+		f.readBuffer = make([]byte, head.length)
+		f.rbuf = f.readBuffer
+	}
+
+	// assume the underlying reader takes care of timeouts and retries
+	n, err := io.ReadFull(f.r, f.rbuf)
+	if err != nil {
+		return fmt.Errorf("unable to read frame body: read %d/%d bytes: %v", n, head.length, err)
+	}
```

Comment:
`io.ReadFull` becomes obsolete if `f.rbuf` is initialized with buffer from `p.buf`.